server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -232,11 +243,17 @@ public InternalEngine(EngineConfig engineConfig) {
             for (ReferenceManager.RefreshListener listener: engineConfig.getInternalRefreshListener()) {
                 this.internalSearcherManager.addListener(listener);
             }
-            this.localCheckpointTracker = createLocalCheckpointTracker(engineConfig, lastCommittedSegmentInfos, logger,
-                () -> acquireSearcher("create_local_checkpoint_tracker", SearcherScope.INTERNAL), localCheckpointTrackerSupplier);
             this.lastRefreshedCheckpointListener = new LastRefreshedCheckpointListener(localCheckpointTracker.getCheckpoint());
             this.internalSearcherManager.addListener(lastRefreshedCheckpointListener);
             maxSeqNoOfUpdatesOrDeletes = new AtomicLong(SequenceNumbers.max(localCheckpointTracker.getMaxSeqNo(), translog.getMaxSeqNo()));
+            if (softDeleteEnabled && localCheckpointTracker.getCheckpoint() < localCheckpointTracker.getMaxSeqNo()) {
+                try (Searcher searcher = acquireSearcher("restore_version_map_and_checkpoint_tracker", SearcherScope.INTERNAL)) {
+                    restoreVersionMapAndCheckpointTracker(Lucene.wrapAllDocsLive(searcher.getDirectoryReader()));
```

Comment:
can we actually call flush after that to make sure we commit and clear the version map again?

Reply:
can we actually skip this entire dance if we don't have any deletes in the index? just curious...

Reply:
> can we actually skip this entire dance if we don't have any deletes in the index? just curious...

Yes, we can. If there's no delete, the version map will empty after this method. For the local checkpoint, I think we would like to have it in sync Lucene index. Yannick, WDYT?

Reply:
> can we actually skip this entire dance if we don't have any deletes in the index? just curious...

I think we could even skip it if there are no soft-deletes in the range we're looking for (i.e. from local checkpoint of safe commit upwards). I would like to avoid special-casing this too much though,  and think we should just always do the full thing here. This should be very fast, given that this is a subset of the range that we will have to replay from the translog, which will certainly take much longer.